### PR TITLE
bochs: configure it via module system

### PR DESCRIPTION
**Pull request by AndersonTorres**
## Description of changes

Playing a little bit with the module system proposed by @Atemu at #312432.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/bo/bochs/options.nix [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  lib,
+  config,
+  stdenv,
+  ...
+}:
+
+{
+  options = {
+    interfaces = lib.mkOption {
+      default = [ "SDL2" "term" "wx" "x11" ];
```
Comment (Atemu) [inline pkgs/by-name/bo/bochs/options.nix:11]: Lists compose quite poorly. A better design could be to make this an attrset or individual options such that you can set them like this:

```Nix
{
  sdl2 = true;
  term = true;
  wx = true;
  x11 = true;
  somethinElse = false;
}
```

Comment (AndersonTorres) [reply pkgs/by-name/bo/bochs/options.nix:11]: > a bunch of individual options

Like the previous commit?
https://github.com/NixOS/nixpkgs/pull/342188/commits/0e3afc49b1f4a4bb24235fa3f3c14d7b905b3005

> a freeform attrset

Like this?

```nix
options = {
  interfaces = lib.mkOption {
    default = {
      sdl =  true;
      term =  true;
      # . . .
    };
  };
};
```

**Comment by Atemu** [comment]
When you have individual options, you could also experiment with having the buildInputs deps declared with each option like I did with ffmpeg.

File: pkgs/by-name/bo/bochs/options.nix (hunk 1 above)
Comment (AndersonTorres) [reply pkgs/by-name/bo/bochs/options.nix]: I like to use Bochs expression as a "test+documentation" example.
I wanna try that settings-like approach just for the docs.

**Comment by Atemu** [comment]
Re: availableOn

You can take this one step further if you have the deps declared with the option and automatically apply it to all packages and decide the default on that.

I had considered doing something like this for ffmpeg too but IIRC it was too costly in eval time for questionable benefit.

I had even gone one step further and tryEval'd the toString of the drv in order to catch any sort of eval failure, including transitive ones.

**Comment by Atemu** [comment]
https://github.com/Atemu/nixpkgs/commit/9a0d986bc6f913a5dfeaf859058a3c0d576c4ecf

**Comment by AndersonTorres** [comment]
I am thinking on how to "extend" this correctly.

Usually, SDL packages need to be put in both buildInputs and nativeBuildInputs, especially with strictDeps set as true.

This way, a `config.sdl = true;` should come with something filling both *lists at the same time.

However
- I don't know if splicing will break (no one does)
- I am unable to remember if buildInputs receive either `pkgs.sdl` or  `buildPackages.sdl`...

**Comment by Atemu** [comment]
> SDL packages need to be put in both buildInputs and nativeBuildInputs, especially with strictDeps set as true.

Would be interesting to know why that is; which executable of SDL gets executed here.

This edge-case can be handled by bespoke logic though while everything else still gets handled generically.

> I don't know if splicing will break (no one does)

I don't see how this would break splicing? The pkgs in `packages` are still spliced; nothing un-spliced them.

> I am unable to remember if buildInputs receive either `pkgs.sdl` or `buildPackages.sdl`...

buildInputs are depsHostTarget but we don't call them that for historical reasons.

buildPackages are pkgsBuildHost but we don't call them that for ??? reasons.

Therefore, buildPackages don't go into buildInputs. (pkgs == pkgsHostTarget)

**Comment by AndersonTorres** [comment]
> Would be interesting to know why that is; which executable of SDL gets executed here.

My guess:

SDL has a `sdl-config`, its own `pkg-config`-like utility. This utility goes into its own `bin` output.
Because of this, packages linking to SDL usually needs SDL in buildInputs (libraries) and nativeBuildInputs too (the config utility).
Otherwise, errors like `sdl-config not found` happen.

On the other hand, Bochs uses `pkg-config` directly, so no problem here.

**Pull request closed without merging**